### ISSUE TITLE
Source domain training and w/o ada performance

### ISSUE THREAD

**Issue by YYinn** [CSCYQJ/MICCAI23-ProtoContra-SFDA#11]
Hi! Thank you for the author's sharing! 

I met some problem in training source domain. The model of ct2mr and mr2ct w/o adaptation is 0.68 and 0.49 for me, which is slightly different from the paper result. Could there be any other possible reason causing the results? I followed the instruction of preprocessing in the paper.

Thank you!

**Comment by YYinn** [comment]
I've got 0.49 as w/o ada for mr2ct and the result of PFA was 0.7153. As i moved on to the CL stage, it went down to 0.1472. As for ct2mr, the result i've got was 0.8262, and the result after CL stage was 0.2548. Could there be any other possible reason causing the results? Thanks!!

**Comment by BarY7** [comment]
Hey @YYinn , did you figure it out? 

In my case the no-adaptation results seem to also fluctuate. The drop in the CL didn't happen in my case though, but the model's total accuracy is about 82.6 at best.

**Comment by YYinn** [comment]
> Hey @YYinn , did you figure it out?
> 
> In my case the no-adaptation results seem to also fluctuate. The drop in the CL didn't happen in my case though, but the model's total accuracy is about 82.6 at best.

hey, @BarY7 , I only achieved result around 0.83 in the direction of CT to MR on the abdominal multi-organ dataset. I think it is related to the hyper-parameters and the performance of the source domain model. If the hyperparameters are carefully adjusted, there may be an improvement in performance.

**Comment by BarY7** [comment]
> > Hey @YYinn , did you figure it out?
> > In my case the no-adaptation results seem to also fluctuate. The drop in the CL didn't happen in my case though, but the model's total accuracy is about 82.6 at best.
> 
> hey, @BarY7 , I only achieved result around 0.83 in the direction of CT to MR on the abdominal multi-organ dataset. I think it is related to the hyper-parameters and the performance of the source domain model. If the hyperparameters are carefully adjusted, there may be an improvement in performance.

Thanks for the response
It seems like the LR in the uploaded config is different than the paper (10^-3 vs. 10^-4), but changing it didn't help either.
I don't think its hyperparams related, as It is weird if the authors uploaded config that doesn't provide similar results to those in the paper. Maybe has something to do with the preprocessing

**Comment by YYinn** [comment]
> > > Hey @YYinn , did you figure it out?
> > > In my case the no-adaptation results seem to also fluctuate. The drop in the CL didn't happen in my case though, but the model's total accuracy is about 82.6 at best.
> > 
> > 
> > hey, @BarY7 , I only achieved result around 0.83 in the direction of CT to MR on the abdominal multi-organ dataset. I think it is related to the hyper-parameters and the performance of the source domain model. If the hyperparameters are carefully adjusted, there may be an improvement in performance.
> 
> Thanks for the response It seems like the LR in the uploaded config is different than the paper (10^-3 vs. 10^-4), but changing it didn't help either. I don't think its hyperparams related, as It is weird if the authors uploaded config that doesn't provide similar results to those in the paper. Maybe has something to do with the preprocessing

Indeed, if you are using parameters from the paper or code, you may need to maintain the same preprocessing process as the author. Similar issues have been mentioned in other issues, you can take a look to resolve the preprocessing problem.

**Comment by qwerasdzxcvb** [comment]
> I've got 0.49 as w/o ada for mr2ct and the result of PFA was 0.7153. As i moved on to the CL stage, it went down to 0.1472. As for ct2mr, the result i've got was 0.8262, and the result after CL stage was 0.2548. Could there be any other possible reason causing the results? Thanks!!
请问你有解决这个问题吗，是调整了什么超参数吗，目前我mrtoct方向的PFA阶段只有0.76，在cttomr方向的PFA阶段有0.83
我的数据预处理也是用的作者的

**Comment by YYinn** [comment]
> > I've got 0.49 as w/o ada for mr2ct and the result of PFA was 0.7153. As i moved on to the CL stage, it went down to 0.1472. As for ct2mr, the result i've got was 0.8262, and the result after CL stage was 0.2548. Could there be any other possible reason causing the results? Thanks!!
> > 请问你有解决这个问题吗，是调整了什么超参数吗，目前我mrtoct方向的PFA阶段只有0.76，在cttomr方向的PFA阶段有0.83
> > 我的数据预处理也是用的作者的

@qwerasdzxcvb @BarY7 
Hello, 
Apart from the influence of preprocessing, I previously adjusted the nav_t parameter, which might have helped to improve the results. But the final results I got were also similar, as 0.8382 and 0.7544, respectively. 
A influencing factor I thought is the "crop out the non-body region" operation mentioned in the paper. It seems that there is no code for this operation on GitHub. If larger areas are manually cropped (e.g. 20 for each side), the results might improve.

**Comment by BarY7** [comment]
@YYinn 
Did you run PFA on the best model achieved in the source training phase?
In my case, the PFA results are very unstable, sometimes they crash the dice to 60%.

Regarding the preprocessing - there is indeed a notebook the authors uploaded with the pre-processing code, although it is a bit messy (some things had to be uncommented). It can be found in the commit history (it was deleted)

"crop out the non-body region"  - For MRI the voxels are thresholded with value 100 and everything below that is considered "non-body" and cropped out.
For ct images is done by the function `delete_bed_torch` - it is uploaded alongside the notebook.

**Comment by BarY7** [comment]
In the CT -> MR direction I did manage to get one model to 87.5% after the CL stage, but when trying a slightly different base model (from a few epochs back for the source model, with about the same results) I get 60% after PFA. I don't understand why the process is so unstable

**Comment by YYinn** [comment]
@BarY7 
Thank you for your explanation! I indeed didn't see this preprocessing code before, I just referred to the reference and author's instructions, and didn't see any related code and functions.

In the CT to MR direction experiment, I also had a result of 0.87 before, but at that time I cropped 10 pixels from the top, bottom, left, and right of the image (to perform the "crop out the non-body region" operation). If I didn't use this operation and didn't modify the hyper-parameters, the result was only 0.83. (Has the delete_bed_torch function been deleted?）

I used the best model for PFA stage training because I saw that this was also done in the config file of the code. I haven't tried using the results of other models, but mainly adjusted nav_t. While during the experiment, I also found that the fluctuation of the model's results was greatly affected by other parameters. 

I have asked in another issue that why my results are different from those in the paper, and the author said that the results of the PFA stage are greatly influenced by the source domain model performance. However, if you are using an source domain model with similar results to the best one, I indeed cannot provide an answer. The model might be quite sensitive to hyper-parameters. Besides, I think it might because of the differences in the preprocessing stage. But without a clear version of the code for preprocessing, it's hard to verify the conjecture.

**Comment by qwerasdzxcvb** [comment]
@BarY7 @YYinn 
Thanks for the answer! I don't think I have "crop out the non-body region", I'll try again!

**Comment by BarY7** [comment]
@YYinn 
I don't think cropping 10px of each side should be enough, the CT images are very weird sometimes with centering, see the example. 

![Screenshot 2024-05-23 155119](https://github.com/CSCYQJ/MICCAI23-ProtoContra-SFDA/assets/26419740/e84a83e9-306b-435e-93e6-1c95bbfce89c)

delete_bed_torch is in another file not in the notebook, previously I did something similar before I found this function using skimage label function

Why did you choose nav_t for adjusting? What is the range you tried with?

With regards to the chosen source model, the source models have similar scores on the source domain, but their result on the target domain varies. I wonder if the authors chose a model with naturally good performance on the target or simply the best one.

**Comment by YYinn** [comment]
@BarY7 
Previously, since there were no specific parameters for cropping, I mainly removed the blank background area (non-body region), so I set the crop value to 10. 

nav_t is a major hyperparameter in ProtoLoss, and I noticed that there are some scale differences in the t2p and p2t loss functions, so I tried adjusting the nav_t. The adjustments were mainly made between 0.1 and 1.5. 

Regarding the choice of the source domain model in the PFA stage, I think your point makes sense.

**Comment by BarY7** [comment]
@YYinn 
I see, I will look into updating nav_t

@YYinn @qwerasdzxcvb 
I do not have access to baidu, but it might be a good idea to download the uploaded model from here: 
https://github.com/CSCYQJ/MICCAI23-ProtoContra-SFDA/issues/15

And check its performance on the preprocessed data

**Comment by YYinn** [comment]
@BarY7 
Thanks! I'll try it. My 'w/o adaptation' results are indeed lower than that in the paper.

**Comment by YYinn** [comment]
@BarY7 

> @YYinn I see, I will look into updating nav_t
> 
> @YYinn @qwerasdzxcvb I do not have access to baidu, but it might be a good idea to download the uploaded model from here: #15
> 
> And check its performance on the preprocessed data

If you need the checkpoint, I can send it to your email.

**Comment by BarY7** [comment]
@YYinn Yeah, It would be great!

Please send it to renaiseth at gmail.com, thanks!

**Comment by YYinn** [comment]
@BarY7 
I have already sent the email, see if you have received it!

**Comment by BarY7** [comment]
@YYinn Got it, thank you 🥇

**Comment by dogeONE-bit** [comment]
@BarY7 您好，请问你经过PFA阶段之后的模型是怎么选取的，我看到作者是每5个step就会在测试集上测试一次，但是直接选取在测试集上最好的结果去做CL阶段应该是不合理的吧。

**Comment by WANGSIQII** [comment]
@BarY7 您好，我想麻烦问下图像预处理的过程，我按照作者提交的历史记录中找到了图像预处理文件，但是按照作者提供预处理后的.npz，执行PFA的Dice结果只有9%，我不清楚是哪里搞错了，所以，能麻烦您给我一份预处理后的数据吗，非常感谢！这是我的邮箱：790089478@qq.com

**Comment by WANGSIQII** [comment]
@YYinn 您好我想麻烦问下图像预处理的过程，我按照作者提交的历史记录中找到了图像预处理文件，但是按照作者提供预处理后的.npz，执行PFA的Dice结果只有9%左右，我不清楚是哪里搞错了，所以，能麻烦您给我一份预处理后的数据吗，非常感谢！这是我的邮箱：[790089478@qq.com](mailto:790089478@qq.com)

**Comment by YYinn** [comment]
@WANGSIQII hihi，我当时没有找到这篇文章的预处理代码，所以复现的时候没有按照他的预处理过程来做。虽然最后结果和原文不同，但是也没有很低，我想应该是正常的。你的源域模型结果正常吗？

**Comment by WANGSIQII** [comment]
你好，感谢您的回复，我的源域模型结果是正常的，大概在90%左右，但是在执行PFA时候就很低，能方便问下有什么调参技巧吗


学生王思齐
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "CSCYQJ/MICCAI23-ProtoContra-SFDA"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年12月31日(星期二) 下午4:45
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [CSCYQJ/MICCAI23-ProtoContra-SFDA] Source domain training and w/o ada performance (Issue #11)





 
@WANGSIQII hihi，我当时没有找到这篇文章的预处理代码，所以复现的时候没有按照他的预处理过程来做。虽然最后结果和原文不同，但是也没有很低，我想应该是正常的。你的源域模型结果正常吗？

**Comment by YYinn** [comment]
@WANGSIQII 你好，我当时调参调了一下nav_t和beta应该是。
如果源域正常的话，或许你可以检查一下源域和目标域图像灰度值会不会没有归到同一个范围呢？

**Comment by WANGSIQII** [comment]
好的，谢谢您的回复，我之前的做法是对于每个域内进行了各自归一化，没有进行跨域的对齐，还想麻烦问下您，归到同一个范围是指将源域中的灰度值向目标域对齐这种吗？感谢您的回复，顺祝您新年快乐！

学生王思齐
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "CSCYQJ/MICCAI23-ProtoContra-SFDA"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年12月31日(星期二) 下午4:58
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [CSCYQJ/MICCAI23-ProtoContra-SFDA] Source domain training and w/o ada performance (Issue #11)





 
@WANGSIQII 你好，我当时调参调了一下nav_t和beta应该是。
 如果源域正常的话，或许你可以检查一下源域和目标域图像灰度值会不会没有归到同一个范围呢？

**Comment by YYinn** [comment]
@WANGSIQII 新年快乐！
我之前说的灰度归一就是在各个域内做的z-score或者是基于最大值最小值的归一化，应该跟你做的是一样的。因为感觉如果源域正常但PFA后结果不正常的话，可能是两个域数据范围不同（比如目标域没有做归一化），或者是PFA代码的问题。

如果PFA确实load 源域模型，但是结果很低的话，我能想到的可能是，可以再看看是不是PFA config中的参数有问题，或者看看loss打印出来是否正常。
祝试验顺利！

**Comment by WANGSIQII** [comment]
好的，感谢您的回复，我再试一试！


学生王思齐
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "CSCYQJ/MICCAI23-ProtoContra-SFDA"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年1月2日(星期四) 中午1:01
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [CSCYQJ/MICCAI23-ProtoContra-SFDA] Source domain training and w/o ada performance (Issue #11)





 
@WANGSIQII 新年快乐！
 我之前说的灰度归一就是在各个域内做的z-score或者是基于最大值最小值的归一化，应该跟你做的是一样的。因为感觉如果源域正常但PFA后结果不正常的话，可能是两个域数据范围不同（比如目标域没有做归一化），或者是PFA代码的问题。
 
如果PFA确实load 源域模型，但是结果很低的话，我能想到的可能是，可以再看看是不是PFA config中的参数有问题，或者看看loss打印出来是否正常。
 祝试验顺利！
